Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs

```diff
@@ -1,39 +1,251 @@
+
+        public void SetupFromInteractionSource(InteractionSourceState interactionSourceState)
+        {
+            //Update the Tracked state of the controller
+            UpdateControllerData(interactionSourceState);
+
+            MixedRealityControllerMapping controllerMapping = Managers.MixedRealityManager.Instance.ActiveProfile.GetControllerMapping(typeof(WindowsMixedRealityController), ControllerHandedness);
+            //MixedRealityControllerMappingProfile controllerMapping = Managers.MixedRealityManager.Instance.ActiveProfile.GetControllerMapping<WindowsMixedRealityController>(ControllerHandedness);
+            if (controllerMapping.Interactions?.Length > 0)
+            {
+                SetupFromMapping(controllerMapping.Interactions);
+            }
+            else
+            {
+                SetupWMRControllerDefaults(interactionSourceState);
+            }
+        }
+
+        private void SetupFromMapping(InteractionMapping[] mappings)
+        {
+            for (uint i = 0; i < mappings.Length; i++)
+            {
+                // Add interaction for Mapping
+                Interactions.Add(mappings[i].InputType, new InteractionMapping(i, mappings[i].AxisType, mappings[i].InputType, mappings[i].InputAction));
+            }
+        }
+
+        private void SetupWMRControllerDefaults(InteractionSourceState interactionSourceState)
+        {
+            InputAction[] inputActions = Managers.MixedRealityManager.Instance.ActiveProfile.InputActionsProfile.InputActions;
+            if (inputActions == null)
+            {
+                return;
+            }
+            //Add the Controller Pointer
+            Interactions.Add(DeviceInputType.SpatialPointer, new InteractionMapping(1, AxisType.SixDoF, DeviceInputType.SpatialPointer, inputActions.GetActionByName("Select"))); // Note will convert these lookups to indexes
+
+            // Add the Controller trigger
+            Interactions.Add(DeviceInputType.Trigger, new InteractionMapping(2, AxisType.SingleAxis, DeviceInputType.Trigger, inputActions.GetActionByName("Select")));
+
+            // If the controller has a Grip / Grasp button, add it to the controller capabilities
+            if (interactionSourceState.source.supportsGrasp)
+            {
+                Interactions.Add(DeviceInputType.SpatialGrip, new InteractionMapping(3, AxisType.SixDoF, DeviceInputType.SpatialGrip, inputActions.GetActionByName("Grip")));
+
+                Interactions.Add(DeviceInputType.GripPress, new InteractionMapping(4, AxisType.SingleAxis, DeviceInputType.GripPress, inputActions.GetActionByName("Grab")));
+            }
+
+            // If the controller has a menu button, add it to the controller capabilities
+            if (interactionSourceState.source.supportsMenu)
+            {
+                Interactions.Add(DeviceInputType.Menu, new InteractionMapping(5, AxisType.Digital, DeviceInputType.Menu, inputActions.GetActionByName("Menu")));
```

Comment:
It feels really odd to have a button described as an axis. I can understand the trigger as it could tell you how far it was depressed, but a basic button?

Reply:
It follows Unity's terminology, where they're all "input axes", but the type can be different things:
![image](https://user-images.githubusercontent.com/3580640/40943567-74822572-6806-11e8-9a14-a21c871ad0e5.png)

Reply:
Axis is correct in my understanding @davidkline-ms , as it describes the type of input data, from Axis 1d through to Axis 6d.  Digital is the only exception as it's an on/off state, pressed or not.

We can only go so far down the generic route when describing these things.

Reply:
I can live with it, then :)

Reply:
Goes back to that essence of input I talked about.

Technically digital isn't an axis per se but it's still the base and subset of all the things you can do with a signal. Essentially digital is just a clamped sinusoidal with a ceiling and floor value.